### ISSUE TITLE
tf.import_graph_def() restricts the order of nodes in graph proto.

### ISSUE THREAD

**Issue by sj6077** [tensorflow/tensorflow#9906]
-----------------------
### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**: Yes
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**:
Linux Ubuntu 14.04
- **TensorFlow installed from (source or binary)**:
source
- **TensorFlow version (use command below)**:
commit 3bee923c9
- **Bazel version (if compiling from source)**:
0.4.5
- **CUDA/cuDNN version**:
cuda 8.0/cudnn 5.1.5
- **GPU model and memory**:
Tesla P40 
- **Exact command to reproduce**:

### Describe the problem
I used auto parallel optimizer in grappler for cifar10 example, but grappler changes the order of nodes in graph proto. It causes the failure of import_graph_def, because ops.set_shapes_for_outputs(op), which is line 407 of tensorflow/tensorflow/python/framework/importer.py is failed. This is because all of the input nodes must be defined before to define a node. importer restricts the order of nodes for graph definition proto, but I think importer shoud be flexible to the order of nodes. 

### Source code / logs
  ```
with tf.Graph().as_default() as graph:
    #global_step = tf.contrib.framework.get_or_create_global_step()
    global_step = tf.get_variable(
        'global_step', [], dtype=tf.int32,
        initializer=tf.constant_initializer(0), trainable=False)

    # Get images and labels for CIFAR-10.
    images, labels = cifar10.distorted_inputs()

    # Build a Graph that computes the logits predictions from the
    # inference model.
    logits = cifar10.inference(images)

    # Calculate loss.
    loss = cifar10.loss(logits, labels)

    # Build a Graph that trains the model with one batch of examples and
    # updates the model parameters.
    train_op = cifar10.train(loss, global_step)

    init_op = tf.global_variables_initializer()

    queue_runners = []
    for qr in ops.get_collection(ops.GraphKeys.QUEUE_RUNNERS):
      queue_runners.append(qr.to_proto())

    mg = meta_graph.create_meta_graph_def(graph=graph)

  #Auto-parallel
  rewriter_config = rewriter_config_pb2.RewriterConfig()
  rewriter_config.optimizers.append('autoparallel')
  rewriter_config.auto_parallel.num_replicas = FLAGS.num_gpus

  graph_def = tf_optimizer.OptimizeGraph(rewriter_config, mg)

  with tf.Graph().as_default() as g:
    tf.import_graph_def(graph_def=graph_def, name='')
```

This is the log
```
 File "/cmsdata/ssd0/soojeong/tensorflow_venv/local/lib/python2.7/site-packages/tensorflow/python/platform/app.py", line 48, in run
    _sys.exit(main(_sys.argv[:1] + flags_passthrough))
  File "tensorflow_rdag/examples/distributed/cifar10/cifar10_in_graph_auto_parallel.py", line 157, in main
    train()
  File "tensorflow_rdag/examples/distributed/cifar10/cifar10_in_graph_auto_parallel.py", line 105, in train
    tf.import_graph_def(graph_def=graph_def, name='')
  File "/cmsdata/ssd0/soojeong/tensorflow_venv/local/lib/python2.7/site-packages/tensorflow/python/framework/importer.py", line 404, in import_graph_def
    ops.set_shapes_for_outputs(op)
  File "/cmsdata/ssd0/soojeong/tensorflow_venv/local/lib/python2.7/site-packages/tensorflow/python/framework/ops.py", line 1723, in set_shapes_for_outputs
    shapes = shape_func(op)
  File "/cmsdata/ssd0/soojeong/tensorflow_venv/local/lib/python2.7/site-packages/tensorflow/python/framework/ops.py", line 1673, in call_with_requiring
    return call_cpp_shape_fn(op, require_shape_fn=True)
  File "/cmsdata/ssd0/soojeong/tensorflow_venv/local/lib/python2.7/site-packages/tensorflow/python/framework/common_shapes.py", line 610, in call_cpp_shape_fn
    debug_python_shape_fn, require_shape_fn)
  File "/cmsdata/ssd0/soojeong/tensorflow_venv/local/lib/python2.7/site-packages/tensorflow/python/framework/common_shapes.py", line 653, in _call_cpp_shape_fn_impl
    v = tensor_util.constant_value(op.inputs[idx])
  File "/cmsdata/ssd0/soojeong/tensorflow_venv/local/lib/python2.7/site-packages/tensorflow/python/framework/tensor_util.py", line 710, in constant_value
    ret = _ConstantValue(tensor)
  File "/cmsdata/ssd0/soojeong/tensorflow_venv/local/lib/python2.7/site-packages/tensorflow/python/framework/tensor_util.py", line 676, in _ConstantValue
    fill_value = constant_value(tensor.op.inputs[1])
  File "/cmsdata/ssd0/soojeong/tensorflow_venv/local/lib/python2.7/site-packages/tensorflow/python/framework/ops.py", line 1443, in __getitem__
    return self._op._inputs[i]
IndexError: list index out of range

```

**Comment by concretevitamin** [comment]
@benoitsteiner: mind taking a look?  Do we have plans to modify `tf.import_graph_def()` for the desired effects?

**Comment by benoitsteiner** [comment]
The easiest solution is probably to update the grappler optimizers to ensure that the final output is generated in topological order. The one difficulty is going to figure out where to break looks to make TensorFlow happy.  The natural place would be the NextIteration node.

**Comment by zhangyaobit** [comment]
@sj6077, it's great/exciting to see you are using this feature! 

While we are still actively developing it and improving its usability/scalability, I hit the same order problem you described when applying it to the RNN PTB model. I'm working on a fix, which will be submitted soon.

We'd be glad to hear more of your feedback and suggestions later.

**Comment by sj6077** [comment]
Thank you. I'll look forward upgraded grappler optimizers!

**Comment by zhangyaobit** [comment]
The change is in. Please give it a try. https://github.com/tensorflow/tensorflow/commit/194b1644cba446d36c6d192771b11245847db7cc

**Comment by sj6077** [comment]
I tried, but there is another issue. I copied cifar10 model, but it doesn't utilize all the gpus. For example, conv2D is not copied for all gpus.

**Comment by zhangyaobit** [comment]
All transitive fanin nodes of target nodes should be copied/replicated for all GPUs. I think Conv2D is very likely to be in the transitive fanin and therefore should be replicated. Could you give more details on this and debug the AutoParallel code a bit to find out why it is not replicated?

I'm aware of the simple_placer issue and will work on a fix. For now, a temporary workaround is
config_proto = tf.ConfigProto(allow_soft_placement=True)

Yes, I will submit an example on ptb RNN model soon.

**Comment by sj6077** [comment]
I'm not sure about Conv2D issue now, I'll figure out it more. I'm now trying to initialize variables, but I couldn't find an easy way. I'll wait for your example.

**Comment by zhangyaobit** [comment]
These two Assign nodes could be initializers for two variables. There are two collections in MetaGraphDef, "trainable_variables" and "variables", where you have a list of VariableDef including a initialization node. After you import the MetaGraphDef, these two collections should be preserved and you can run all initialization nodes in VariableDef.

Could you be more specific about why queue_running doesn't work with MonitoredTrainingSession? Looks like MonitoredTrainingSession uses queue runner. https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/training/monitored_session.py#L541

Yes, follow-up work will support multiple machines with both synchronous and async training; it may take some time for us to get there. For now, I'm not sure if using AutoParallel within a machine on multiple devices could be combined with a machine-level parallelization with SyncReplicasOptimizer; looks like they two should be compatible.

**Comment by sj6077** [comment]
For the queue runner, I think it's same as initializer. grappler returns graph instead of meta graph, so I can gather only subset of initializers from old meta graph. 

` with tf.Graph().as_default():
    meta_graph_def = meta_graph_pb2.MetaGraphDef()
    meta_graph_def.ParseFromString(meta_graph_def_str)
    meta_graph.rdag_import_meta_graph(meta_graph_def)

    rewriter_config = rewriter_config_pb2.RewriterConfig()
    rewriter_config.optimizers.append('autoparallel')
    rewriter_config.auto_parallel.num_replicas = num_replicas

    graph_def = tf_optimizer.OptimizeGraph(rewriter_config, meta_graph_def)
    importer.rdag_import_graph_def(graph_def=graph_def, name='')

    meta_graph_def = tf.train.export_meta_graph()`


My initializer nodes contains below, but actually more variables to be initialized by with Assign_1 such as `local3/weights/ExponentialMovingAverage/Assign_1'. I got init_op from tf.global_variables_initializer(), and MonitoredTrainingSession also use same initializer for this, but the problem is it only half contains initializers for all the variables.

`op: "NoOp"
input: "^global_step/Assign"
input: "^conv1/weights/Assign"
input: "^conv1/biases/Assign"
input: "^conv2/weights/Assign"
input: "^conv2/biases/Assign"
input: "^local3/weights/Assign"
input: "^local3/biases/Assign"
input: "^local4/weights/Assign"
input: "^local4/biases/Assign"
input: "^softmax_linear/weights/Assign"
input: "^softmax_linear/biases/Assign"
input: "^conv1/weight_loss/avg/Assign"
input: "^conv2/weight_loss/avg/Assign"
input: "^local3/weight_loss/avg/Assign"
input: "^local4/weight_loss/avg/Assign"
input: "^softmax_linear/weight_loss/avg/Assign"
input: "^cross_entropy/avg/Assign"
input: "^total_loss/avg/Assign"
input: "^conv1/weights/ExponentialMovingAverage/Assign"
input: "^conv1/biases/ExponentialMovingAverage/Assign"
input: "^conv2/weights/ExponentialMovingAverage/Assign"
input: "^conv2/biases/ExponentialMovingAverage/Assign"
input: "^local3/weights/ExponentialMovingAverage/Assign"
input: "^local3/biases/ExponentialMovingAverage/Assign"
input: "^local4/weights/ExponentialMovingAverage/Assign"
input: "^local4/biases/ExponentialMovingAverage/Assign"
input: "^softmax_linear/weights/ExponentialMovingAverage/Assign"
input: "^softmax_linear/biases/ExponentialMovingAverage/Assign"
device: "/device:CPU:0"
`

**Comment by sj6077** [comment]
One more question about "trainable_variables" and "variables". Do you copy trainable_variables as variables, and each gpu uptdates variables independently, and finally updates trainable_variables?

**Comment by zhangyaobit** [comment]
Yes, the collections are not updated and thus need manually updated currently. This is cumbersome and we will definitely think of ways to improve (will first get a working, proof-of-concept ptb example, and then improve/streamline the usage). In particular, you are exactly right that the snapshot name in VariableDef needs to be updated. For ptb, the names of all tensors (the states) that the model refer need to be updated (see the code below).

I will just post the ptb_word_lm.py   code here (before I submit it) for your reference. Could you try follow its usage pattern used here?

`"""Example / benchmark for building a PTB LSTM model.

Trains the model described in:
(Zaremba, et. al.) Recurrent Neural Network Regularization
http://arxiv.org/abs/1409.2329

There are 3 supported model configurations:
===========================================
| config | epochs | train | valid  | test
===========================================
| small  | 13     | 37.99 | 121.39 | 115.91
| medium | 39     | 48.45 |  86.16 |  82.07
| large  | 55     | 37.87 |  82.62 |  78.29
The exact results may vary depending on the random initialization.

The hyperparameters used in the model:
- init_scale - the initial scale of the weights
- learning_rate - the initial value of the learning rate
- max_grad_norm - the maximum permissible norm of the gradient
- num_layers - the number of LSTM layers
- num_steps - the number of unrolled steps of LSTM
- hidden_size - the number of LSTM units
- max_epoch - the number of epochs trained with the initial learning rate
- max_max_epoch - the total number of epochs for training
- keep_prob - the probability of keeping weights in the dropout layer
- lr_decay - the decay of the learning rate for each epoch after "max_epoch"
- batch_size - the batch size

The data required for this example is in the data/ dir of the
PTB dataset from Tomas Mikolov's webpage:

$ wget http://www.fit.vutbr.cz/~imikolov/rnnlm/simple-examples.tgz
$ tar xvf simple-examples.tgz

To run:

$ python ptb_word_lm.py --data_path=simple-examples/data/

"""
from __future__ import absolute_import
from __future__ import division
from __future__ import print_function

import time

import google3
import numpy as np
import tensorflow as tf

from google3.third_party.tensorflow.core.framework import variable_pb2
from google3.third_party.tensorflow.python.grappler import tf_optimizer
from google3.third_party.tensorflow_models.tutorials.rnn.ptb import reader

flags = tf.flags
logging = tf.logging

flags.DEFINE_string(
    "model", "small",
    "A type of model. Possible options are: small, medium, large.")
flags.DEFINE_string("data_path", None,
                    "Where the training/test data is stored.")
flags.DEFINE_string("save_path", None,
                    "Model output directory.")
flags.DEFINE_bool("use_fp16", False,
                  "Train using 16-bit floats instead of 32bit floats")
flags.DEFINE_integer("num_gpus", 1,
                     "If larger than 1, Grappler AutoParallel optimizer "
                     "will create multiple training replicas with each GPU "
                     "running one replica.")

FLAGS = flags.FLAGS


def export_state_tuples(state_tuples, name):
  for state_tuple in state_tuples:
    tf.add_to_collection(name, state_tuple.c)
    tf.add_to_collection(name, state_tuple.h)


def import_state_tuples(state_tuples, name, num_replicas):
  restored = []
  for i in range(len(state_tuples) * num_replicas):
    c = tf.get_collection_ref(name)[2 * i + 0]
    h = tf.get_collection_ref(name)[2 * i + 1]
    restored.append(tf.contrib.rnn.LSTMStateTuple(c, h))
  return tuple(restored)


def with_autoparallel_prefix(replica_id, name):
  return "AutoParallel-Replica-%d/%s" % (replica_id, name)


class UpdateCollection(object):
  """Update collection info in MetaGraphDef for AutoParallel optimizer."""

  def __init__(self, metagraph, model):
    self._metagraph = metagraph
    self.replicate_states(model.initial_state_name)
    self.replicate_states(model.final_state_name)
    self.update_snapshot_name("variables")
    self.update_snapshot_name("trainable_variables")

  def update_snapshot_name(self, var_coll_name):
    var_list = self._metagraph.collection_def[var_coll_name]
    for i, value in enumerate(var_list.bytes_list.value):
      var_def = variable_pb2.VariableDef()
      var_def.ParseFromString(value)
      # Somehow node Model/global_step/read doesn't have any fanout and seems to
      # be only used for snapshot; this is different from all other variables.
      if var_def.snapshot_name != "Model/global_step/read:0":
        var_def.snapshot_name = with_autoparallel_prefix(
            0, var_def.snapshot_name)
      value = var_def.SerializeToString()
      var_list.bytes_list.value[i] = value

  def replicate_states(self, state_coll_name):
    state_list = self._metagraph.collection_def[state_coll_name]
    num_states = len(state_list.node_list.value)
    for replica_id in range(1, FLAGS.num_gpus):
      for i in range(num_states):
        state_list.node_list.value.append(state_list.node_list.value[i])
    for replica_id in range(FLAGS.num_gpus):
      for i in range(num_states):
        index = replica_id * num_states + i
        state_list.node_list.value[index] = with_autoparallel_prefix(
            replica_id, state_list.node_list.value[index])


def data_type():
  return tf.float16 if FLAGS.use_fp16 else tf.float32


class PTBInput(object):
  """The input data."""

  def __init__(self, config, data, name=None):
    self.batch_size = batch_size = config.batch_size
    self.num_steps = num_steps = config.num_steps
    self.epoch_size = ((len(data) // batch_size) - 1) // num_steps
    self.input_data, self.targets = reader.ptb_producer(
        data, batch_size, num_steps, name=name)


class PTBModel(object):
  """The PTB model."""

  def __init__(self, is_training, config, input_):
    self._is_training = is_training
    self._input = input_

    batch_size = input_.batch_size
    num_steps = input_.num_steps
    size = config.hidden_size
    vocab_size = config.vocab_size

    # Slightly better results can be obtained with forget gate biases
    # initialized to 1 but the hyperparameters of the model would need to be
    # different than reported in the paper.
    def lstm_cell():
      return tf.contrib.rnn.BasicLSTMCell(
          size, forget_bias=0.0, state_is_tuple=True, reuse=not is_training)

    attn_cell = lstm_cell
    if is_training and config.keep_prob < 1:
      def attn_cell():
        return tf.contrib.rnn.DropoutWrapper(
            lstm_cell(), output_keep_prob=config.keep_prob)

    cell = tf.contrib.rnn.MultiRNNCell(
        [attn_cell() for _ in range(config.num_layers)], state_is_tuple=True)

    self._initial_state = cell.zero_state(batch_size, data_type())

    with tf.device("/cpu:0"):
      embedding = tf.get_variable(
          "embedding", [vocab_size, size], dtype=data_type())
      inputs = tf.nn.embedding_lookup(embedding, input_.input_data)

    if is_training and config.keep_prob < 1:
      inputs = tf.nn.dropout(inputs, config.keep_prob)

    # Simplified version of tensorflow_models/tutorials/rnn/rnn.py's rnn().
    # This builds an unrolled LSTM for tutorial purposes only.
    # In general, use the rnn() or state_saving_rnn() from rnn.py.
    #
    # The alternative version of the code below is:
    #
    # inputs = tf.unstack(inputs, num=num_steps, axis=1)
    # outputs, state = tf.contrib.rnn.static_rnn(cell, inputs,
    #                            initial_state=self._initial_state)
    outputs = []
    state = self._initial_state
    with tf.variable_scope("RNN"):
      for time_step in range(num_steps):
        if time_step > 0: tf.get_variable_scope().reuse_variables()
        (cell_output, state) = cell(inputs[:, time_step, :], state)
        outputs.append(cell_output)

    output = tf.reshape(tf.concat(outputs, 1), [-1, size])
    softmax_w = tf.get_variable(
        "softmax_w", [size, vocab_size], dtype=data_type())
    softmax_b = tf.get_variable("softmax_b", [vocab_size], dtype=data_type())
    logits = tf.matmul(output, softmax_w) + softmax_b
    loss = tf.contrib.legacy_seq2seq.sequence_loss_by_example(
        [logits],
        [tf.reshape(input_.targets, [-1])],
        [tf.ones([batch_size * num_steps], dtype=data_type())])
    self._cost = cost = tf.reduce_sum(loss) / batch_size
    self._final_state = state

    if not is_training:
      return

    self._lr = tf.Variable(0.0, trainable=False)
    tvars = tf.trainable_variables()
    grads, _ = tf.clip_by_global_norm(tf.gradients(cost, tvars),
                                      config.max_grad_norm)
    optimizer = tf.train.GradientDescentOptimizer(self._lr)
    self._train_op = optimizer.apply_gradients(
        zip(grads, tvars),
        global_step=tf.contrib.framework.get_or_create_global_step())

    self._new_lr = tf.placeholder(
        tf.float32, shape=[], name="new_learning_rate")
    self._lr_update = tf.assign(self._lr, self._new_lr)

  def assign_lr(self, session, lr_value):
    session.run(self._lr_update, feed_dict={self._new_lr: lr_value})

  def export_ops(self, name):
    if self._is_training:
      tf.add_to_collection("lr", self._lr)
      tf.add_to_collection("new_lr", self._new_lr)
      tf.add_to_collection("lr_update", self._lr_update)
    self._name = name
    tf.add_to_collection(self._name + "/cost", self._cost)
    self._initial_state_name = self._name + "/initial"
    self._final_state_name = self._name + "/final"
    export_state_tuples(self._initial_state, self._initial_state_name)
    export_state_tuples(self._final_state, self._final_state_name)

  def import_ops(self):
    if self._is_training:
      self._train_op = tf.get_collection_ref("train_op")[0]
      self._lr = tf.get_collection_ref("lr")[0]
      self._new_lr = tf.get_collection_ref("new_lr")[0]
      self._lr_update = tf.get_collection_ref("lr_update")[0]
    self._cost = tf.get_collection_ref(self._name + "/cost")[0]
    num_replicas = FLAGS.num_gpus if self._name == "Train" else 1
    self._initial_state = import_state_tuples(
        self._initial_state, self._initial_state_name, num_replicas)
    self._final_state = import_state_tuples(
        self._final_state, self._final_state_name, num_replicas)

  @property
  def input(self):
    return self._input

  @property
  def initial_state(self):
    return self._initial_state

  @property
  def cost(self):
    return self._cost

  @property
  def final_state(self):
    return self._final_state

  @property
  def lr(self):
    return self._lr

  @property
  def train_op(self):
    return self._train_op

  @property
  def initial_state_name(self):
    return self._initial_state_name

  @property
  def final_state_name(self):
    return self._final_state_name


class SmallConfig(object):
  """Small config."""
  init_scale = 0.1
  learning_rate = 1.0
  max_grad_norm = 5
  num_layers = 2
  num_steps = 20
  hidden_size = 200
  max_epoch = 4
  max_max_epoch = 13
  keep_prob = 1.0
  lr_decay = 0.5
  batch_size = 20
  vocab_size = 10000


class MediumConfig(object):
  """Medium config."""
  init_scale = 0.05
  learning_rate = 1.0
  max_grad_norm = 5
  num_layers = 2
  num_steps = 35
  hidden_size = 650
  max_epoch = 6
  max_max_epoch = 39
  keep_prob = 0.5
  lr_decay = 0.8
  batch_size = 20
  vocab_size = 10000


class LargeConfig(object):
  """Large config."""
  init_scale = 0.04
  learning_rate = 1.0
  max_grad_norm = 10
  num_layers = 2
  num_steps = 35
  hidden_size = 1500
  max_epoch = 14
  max_max_epoch = 55
  keep_prob = 0.35
  lr_decay = 1 / 1.15
  batch_size = 20
  vocab_size = 10000


class TestConfig(object):
  """Tiny config, for testing."""
  init_scale = 0.1
  learning_rate = 1.0
  max_grad_norm = 1
  num_layers = 1
  num_steps = 2
  hidden_size = 2
  max_epoch = 1
  max_max_epoch = 1
  keep_prob = 1.0
  lr_decay = 0.5
  batch_size = 20
  vocab_size = 10000


def run_epoch(session, model, eval_op=None, verbose=False):
  """Runs the model on the given data."""
  start_time = time.time()
  costs = 0.0
  iters = 0
  state = session.run(model.initial_state)

  fetches = {
      "cost": model.cost,
      "final_state": model.final_state,
  }
  if eval_op is not None:
    fetches["eval_op"] = eval_op

  for step in range(model.input.epoch_size):
    feed_dict = {}
    for i, (c, h) in enumerate(model.initial_state):
      feed_dict[c] = state[i].c
      feed_dict[h] = state[i].h

    vals = session.run(fetches, feed_dict)
    cost = vals["cost"]
    state = vals["final_state"]

    costs += cost
    iters += model.input.num_steps

    if verbose and step % (model.input.epoch_size // 10) == 10:
      print("%.3f perplexity: %.3f speed: %.0f wps" %
            (step * 1.0 / model.input.epoch_size, np.exp(costs / iters),
             iters * model.input.batch_size / (time.time() - start_time)))

  return np.exp(costs / iters)


def get_config():
  if FLAGS.model == "small":
    return SmallConfig()
  elif FLAGS.model == "medium":
    return MediumConfig()
  elif FLAGS.model == "large":
    return LargeConfig()
  elif FLAGS.model == "test":
    return TestConfig()
  else:
    raise ValueError("Invalid model: %s", FLAGS.model)


def main(_):
  if not FLAGS.data_path:
    raise ValueError("Must set --data_path to PTB data directory")

  raw_data = reader.ptb_raw_data(FLAGS.data_path)
  train_data, valid_data, test_data, _ = raw_data

  config = get_config()
  eval_config = get_config()
  eval_config.batch_size = 1
  eval_config.num_steps = 1

  with tf.Graph().as_default():
    initializer = tf.random_uniform_initializer(-config.init_scale,
                                                config.init_scale)

    with tf.name_scope("Train"):
      train_input = PTBInput(config=config, data=train_data, name="TrainInput")
      with tf.variable_scope("Model", reuse=None, initializer=initializer):
        m = PTBModel(is_training=True, config=config, input_=train_input)
        m.export_ops("Train")
      tf.summary.scalar("Training Loss", m.cost)
      tf.summary.scalar("Learning Rate", m.lr)

    with tf.name_scope("Valid"):
      valid_input = PTBInput(config=config, data=valid_data, name="ValidInput")
      with tf.variable_scope("Model", reuse=True, initializer=initializer):
        mvalid = PTBModel(is_training=False, config=config, input_=valid_input)
        mvalid.export_ops("Valid")
      tf.summary.scalar("Validation Loss", mvalid.cost)

    with tf.name_scope("Test"):
      test_input = PTBInput(
          config=eval_config, data=test_data, name="TestInput")
      with tf.variable_scope("Model", reuse=True, initializer=initializer):
        mtest = PTBModel(is_training=False, config=eval_config,
                         input_=test_input)
        mtest.export_ops("Test")

    metagraph = tf.train.export_meta_graph()
    if FLAGS.num_gpus > 1:
      rewriter_config = tf.RewriterConfig()
      rewriter_config.optimizers.append("autoparallel")
      rewriter_config.auto_parallel.enable = True
      rewriter_config.auto_parallel.num_replicas = FLAGS.num_gpus
      optimized_graph = tf_optimizer.OptimizeGraph(rewriter_config, metagraph)
      metagraph.graph_def.CopyFrom(optimized_graph)
      UpdateCollection(metagraph, m)

  with tf.Graph().as_default():
    tf.train.import_meta_graph(metagraph)
    m.import_ops()
    mvalid.import_ops()
    mtest.import_ops()
    sv = tf.train.Supervisor(logdir=FLAGS.save_path)
    config_proto = tf.ConfigProto(allow_soft_placement=True)
    with sv.managed_session(config=config_proto) as session:
      for i in range(config.max_max_epoch):
        lr_decay = config.lr_decay ** max(i + 1 - config.max_epoch, 0.0)
        m.assign_lr(session, config.learning_rate * lr_decay)

        print("Epoch: %d Learning rate: %.3f" % (i + 1, session.run(m.lr)))
        train_perplexity = run_epoch(session, m, eval_op=m.train_op,
                                     verbose=True)
        print("Epoch: %d Train Perplexity: %.3f" % (i + 1, train_perplexity))
        valid_perplexity = run_epoch(session, mvalid)
        print("Epoch: %d Valid Perplexity: %.3f" % (i + 1, valid_perplexity))

      test_perplexity = run_epoch(session, mtest)
      print("Test Perplexity: %.3f" % test_perplexity)

      if FLAGS.save_path:
        print("Saving model to %s." % FLAGS.save_path)
        sv.saver.save(session, FLAGS.save_path, global_step=sv.global_step)


if __name__ == "__main__":
  tf.app.run()`

**Comment by sj6077** [comment]
Thank you for your example! It's really helpful. 

I tested your example, but it has no speed up acrroding to number of gpus increases. Is it just a problem of log?
    ```
if verbose and step % (model.input.epoch_size // 10) == 10:
      print("%.3f perplexity: %.3f speed: %.0f wps" %
            (step * 1.0 / model.input.epoch_size, np.exp(costs / iters),
             iters * model.input.batch_size / (time.time() - start_time)))
```

**Comment by zhangyaobit** [comment]
Yes, I believe it is a problem with logging. If you use tensorboard, you can compare global steps/sec and time to certain accuracy loss.

**Comment by sj6077** [comment]
I found 2 issues related with auto-parallel.
It doesn't replicate queues and euqueue ops, but if input pipeline contains a work for data-preprocessing, gpus are utilized as serial orders.
One more thing is apply gradient descent is a long time task, below is the tracing result, the computation time was 100ms but apply gradient descent time takes 100ms times more. Is there any dpendencies between different apply gradient descent ops for same variables? 
![2017-05-29 16 41 58](https://cloud.githubusercontent.com/assets/2465713/26540740/c1cfef50-448d-11e7-8661-a454bd7d7648.png)

**Comment by zhangyaobit** [comment]
The queue and enqueue ops don't need to be replicated; only dequeue node is replicated, so that multiple GPUs could dequeue and work in parallel. If there is a lot of preprocessing for input data, you may consider using more threads for input preprocessing, so that it keeps up the GPU computations.

The time for apply gradient depends on the models. For small models, it may become a bottleneck; for larger models like the PTB one, it takes relatively a small portion of total time. Could you label the each row in the picture, and mark the related op names?

**Comment by zhangyaobit** [comment]
Closing for now; feel free to reopen if needed.

**Comment by javacop01** [comment]
zhangyaobit,
I know you seem to have fixed this issue with help from sj6077. I am seeing the same issue while running PTB example with tensorflow 1.2.1. Could you please confirm if these changes are already part of 1.2.1 release ? If not, what's the workaround ?

**Comment by zhangyaobit** [comment]
I took a look. This change is not in 1.2.1; it is in 1.3.0. Are you able to use 1.3.0?

What's the error message do you see (could you post the stack trace)?

**Comment by JinScientist** [comment]
Hi zhangyaobit, I have tried with the ptb example with update collection utility,   I still get this error under version 1.4.1
`Traceback (most recent call last):
  File "ppmodel.py", line 558, in <module>
    tf.app.run()
  File "/home/ubuntu/tensorflow/local/lib/python2.7/site-packages/tensorflow/python/platform/app.py", line 48, in run
    _sys.exit(main(_sys.argv[:1] + flags_passthrough))
  File "ppmodel.py", line 524, in main
    tf.train.import_meta_graph(metagraph)
  File "/home/ubuntu/tensorflow/local/lib/python2.7/site-packages/tensorflow/python/training/saver.py", line 1810, in import_meta_graph
    **kwargs)
  File "/home/ubuntu/tensorflow/local/lib/python2.7/site-packages/tensorflow/python/framework/meta_graph.py", line 696, in import_scoped_meta_graph
    ops.prepend_name_scope(value, scope_to_prepend_to_names))
  File "/home/ubuntu/tensorflow/local/lib/python2.7/site-packages/tensorflow/python/framework/ops.py", line 3035, in as_graph_element
    return self._as_graph_element_locked(obj, allow_tensor, allow_operation)
  File "/home/ubuntu/tensorflow/local/lib/python2.7/site-packages/tensorflow/python/framework/ops.py", line 3077, in _as_graph_element_locked
    "graph." % (repr(name), repr(op_name)))
KeyError: "The name 'Train/TrainInput/StridedSlice_1:0' refers to a Tensor which does not exist. The operation, 'Train/TrainInput/StridedSlice_1', does not exist in the graph."`
